### PR TITLE
Verify the IO behavior (a.k.a., basis PR)

### PR DESCRIPTION
**Pull request by Dspil**
This branch contains will be merged into master when the IO behavior is verified and the checks are passing. In the meanwhile, we will use it as the base branch where we merge all the other PRs that have to do with IO verification.

This PR includes changes from #237. Let's all review it and agree on the "base branch" once this PR is merged into master and master is merged here

The assumed lemmas that we depend on for this PR are discussed in #311

### CODE REVIEW

File: router/dataplane.go [hunk 1]
```diff
@@ -897,6 +897,12 @@ func (d *DataPlane) Run(ctx context.Context /*@, ghost place io.Place, ghost sta
 					// error metric
 					continue
 				}
+				// @ ghost *ioSharedArg.State = sN
+				// @ ghost *ioSharedArg.Place = tN
+				// @ MultiElemWitnessConv(ioSharedArg.IBufY, ioIngressID, ioValSeq)
+				// @ fold SharedInv!< dp, ioSharedArg !>()
+				// @ ioLock.Unlock()
+				// End of multi recv event
```
Comment (jcp19) [inline router/dataplane.go:905]:
```suggestion
```

**Review by mlimbeck — APPROVED** [review]

File: router/dataplane.go [hunk 2]
```diff
@@ -1390,12 +1390,22 @@ func (p *scionPacketProcessor) reset() (err error) {
...
 // @ ensures  reserr != nil ==> reserr.ErrorMem()
+// contracts for IO-spec
+// @ requires dp.Valid()
+// @ requires acc(ioLock.LockP(), _) && ioLock.LockInv() == SharedInv!< dp, ioSharedArg !>;
+// @ requires acc(&p.ingressID)
+// @ requires let absPkt := absIO_val(dp, rawPkt, p.ingressID) in
+// @	absPkt.isIO_val_Pkt2 ==> ElemWitness(ioSharedArg.IBufY, ifsToIO_ifs(p.ingressID), absPkt.IO_val_Pkt2_2)
+// @ ensures acc(&p.ingressID)
+// @ ensures reserr == nil && newAbsPkt.isIO_val_Pkt2 ==> ElemWitness(ioSharedArg.OBufY, newAbsPkt.IO_val_Pkt2_1, newAbsPkt.IO_val_Pkt2_2)
+// @ ensures reserr == nil && newAbsPkt.isIO_val_Pkt2 ==> newAbsPkt == absIO_val(dp, respr.OutPkt, respr.EgressID)
```
Comment (jcp19) [inline router/dataplane.go:1401]: we should break these lines to make them easier to read

File: router/dataplane.go [hunk 3]
```diff
@@ -1390,12 +1390,22 @@ func (p *scionPacketProcessor) reset() (err error) {
...
+// TODO: We need to prove that whenever we have a valid scion packet we call processSCION
```
Comment (jcp19) [inline router/dataplane.go:1402]: On a first step, we will prove that whenever we have a valid scion packet in `processSCION`, the correct "next packet" is computed

File: router/dataplane.go [hunk 4]
```diff
@@ -1672,7 +1682,16 @@ func (p *scionPacketProcessor) processIntraBFD(data []byte) (res error) {
...
-func (p *scionPacketProcessor) processSCION( /*@ ghost ub []byte, ghost llIsNil bool, ghost startLL int, ghost endLL int @*/ ) (respr processResult, reserr error /*@ , addrAliasesPkt bool @*/) {
+// contracts for IO-spec
+// @ requires dp.Valid()
+// @ requires acc(ioLock.LockP(), _) && ioLock.LockInv() == SharedInv!< dp, ioSharedArg !>;
+// @ requires acc(&p.ingressID)
+// @ requires let absPkt := absIO_val(dp, p.rawPkt, p.ingressID) in
+// @	absPkt.isIO_val_Pkt2 ==> ElemWitness(ioSharedArg.IBufY, ifsToIO_ifs(p.ingressID), absPkt.IO_val_Pkt2_2)
+// @ ensures acc(&p.ingressID)
+// @ ensures reserr == nil && newAbsPkt.isIO_val_Pkt2 ==> ElemWitness(ioSharedArg.OBufY, newAbsPkt.IO_val_Pkt2_1, newAbsPkt.IO_val_Pkt2_2)
+// @ ensures reserr == nil && newAbsPkt.isIO_val_Pkt2 ==> newAbsPkt == absIO_val(dp, respr.OutPkt, respr.EgressID)
```
Comment (jcp19) [inline router/dataplane.go:1693]: We should again split this to make it easier to read

File: router/dataplane.go (hunk 2 above)
Comment (jcp19) [reply router/dataplane.go:1401]: Additionally, I wonder if `reserr == nil && newAbsPkt.isIO_val_Pkt2 ==> newAbsPkt == absIO_val(dp, respr.OutPkt, respr.EgressID)` should be `reserr == nil ==> newAbsPkt.isIO_val_Pkt2 && newAbsPkt == absIO_val(dp, respr.OutPkt, respr.EgressID)` instead. Shouldn't we always return an `IO_val_Pkt2` when `reserr` is `nil`?

File: router/dataplane.go (hunk 4 above)
Comment (jcp19) [reply router/dataplane.go:1693]: Same comment as before about `newAbsPkt.isIO_val_Pkt2`

File: router/dataplane.go (hunk 2 above)
Comment (mlimbeck) [reply router/dataplane.go:1401]: Not necessarily; The condition `newAbsPkt.isIO_val_Pkt2` is true only when we receive a SCION packet. For instance, if we receive an EPIC, `newAbsPkt.isIO_val_Pkt2` will not be true, even if `reserr` is `nil`.

File: router/dataplane.go (hunk 4 above)
Comment (mlimbeck) [reply router/dataplane.go:1693]: Here I agree with you. I think we should change the precondition.

**Review by jcp19 — APPROVED** [review]

File: pkg/slayers/path/infofield.go [hunk 5]
```diff
@@ -113,10 +132,15 @@ func (inf *InfoField) SerializeTo(b []byte) (err error) {
...
+// requires hf.HVF == AbsMac(hfMac[:])
+// @ ensures absUinfo_(inf.SegID) == old(io.upd_uinfo(absUinfo_(inf.SegID), hf))
 // @ decreases
-func (inf *InfoField) UpdateSegID(hfMac [MacLen]byte) {
+func (inf *InfoField) UpdateSegID(hfMac [MacLen]byte /* @, ghost hf io.IO_HF @ */) {
 	//@ share hfMac
 	inf.SegID = inf.SegID ^ binary.BigEndian.Uint16(hfMac[:2])
+	// TODO: (Markus) How to argue that we need to assume this
+	// upd_uinfo equivalent to xor in the protocol
+	// @ assume absUinfo_(inf.SegID) == old(io.upd_uinfo(absUinfo_(inf.SegID), hf))
```
Comment (jcp19) [inline pkg/slayers/path/infofield.go:143]: This should be an `AssumeForIO`

**Pull request merged**